File: test/python/parse_helper.py
class ParseHelper:
    def assertParse(self, tuple, actual):
        self.assertParseInner(tuple, actual.elements[1])

    def assertParseElements(self, elems, action_args):
        self.assertEqual(5, len(action_args))
        self.assertEqual(len(elems), len(action_args[4]))

        for i, elem in enumerate(elems):
            self.assertParseInner(elem, action_args[4][i])

    def assertParseInner(self, tuple, actual):
        text, offset = tuple[0:2]

        self.assertEqual(text, actual.text)
        self.assertEqual(offset, actual.offset)

        if len(tuple) > 2:
            elements = tuple[2]
            self.assertEqual(len(elements), len(actual.elements))

            for i, elem in enumerate(elements):
                self.assertParseInner(elem, actual.elements[i])

        if len(tuple) > 3:
            labelled = tuple[3]
            for key, value in labelled.items():
                self.assertParseInner(value, getattr(actual, key))

File: test/ruby/parse_helper.rb
module ParseHelper
  def assert_parse(tuple, actual)
    assert_parse_inner(tuple, actual.elements[1])
  end

  def assert_parse_elements(elems, action_args)
    assert_equal 5, action_args.size
    assert_equal elems.size, action_args[4].size

    elems.each_with_index do |elem, i|
      assert_parse_inner(elem, action_args[4][i])
    end
  end

  def assert_parse_inner(tuple, actual)
    text, offset, elements, labelled = tuple

    assert_equal text, actual.text
    assert_equal offset, actual.offset

    if elements
      assert_equal elements.size, actual.elements.size

      elements.each_with_index do |elem, i|
        assert_parse_inner(elem, actual.elements[i])
      end
    end

    if labelled
      labelled.each do |key, value|
        assert_parse_inner(value, actual.__send__(key))
      end
    end
  end
end
